src/core/segment_reader.rs

```diff
@@ -280,6 +282,94 @@ impl SegmentReader {
         Ok(inv_idx_reader)
     }
 
+    /// Returns the list of fields that have been indexed in the segment.
+    /// The field list includes the field defined in the schema as well as the fields
+    /// that have been indexed as a part of a JSON field.
+    /// The returned field name is the full field name, including the name of the JSON field.
+    pub fn fields_metadata(&self) -> crate::Result<Vec<FieldMetadata>> {
+        let mut indexed_fields: Vec<(String, Type)> = Vec::new();
+        for (field, field_entry) in self.schema().fields() {
+            let field_name = field_entry.name().to_string();
+            let is_indexed = field_entry.is_indexed();
+
+            if is_indexed {
+                indexed_fields.push((
+                    field_name.to_string(),
+                    field_entry.field_type().value_type(),
+                ));
+                let is_json = field_entry.field_type().value_type() == Type::Json;
+                if is_json {
+                    let inv_index = self.inverted_index(field)?;
+                    let encoded_fields_in_index = inv_index.list_encoded_fields()?;
+                    indexed_fields.extend(
+                        encoded_fields_in_index
+                            .into_iter()
+                            .map(|(name, typ)| (format!("{}.{}", field_name, name), typ)),
+                    );
+                }
+            }
+        }
+        let mut fast_fields: Vec<(String, Type)> = self
+            .fast_fields()
+            .columnar()
+            .iter_columns()?
+            .map(|(name, handle)| (name.to_string(), Type::from(handle.column_type())))
+            .collect();
+        // Since the type is encoded differently in the fast field and in the inverted index,
+        // the order of the fields is not guaranteed to be the same. Therefore, we sort the fields.
+        // If we are sure that the order is the same, we can remove this sort.
+        indexed_fields.sort_unstable();
+        fast_fields.sort_unstable();
+
+        // Maybe too slow for the high cardinality case
+        fn is_field_stored(field_name: &str, schema: &Schema) -> bool {
```

Comment:
Of course this will be missing some fields that are just STORED and in the dynamic mode, but I think this is the right trade-off!

Reply:
could `Schema::find_field_with_default<'a>` be used? I am a bit scared by spreading the logic in a non-trivial way.
for instance here we don't support escaped "dot".

Reply:
Yes, I'll switch to calling into schema